waspc/src/Wasp/Lib.hs

```diff
@@ -36,52 +37,67 @@ compile ::
   CompileOptions ->
   IO ([CompileWarning], [CompileError])
 compile waspDir outDir options = do
-  appSpecOrCompileErrors <- analyzeWaspProject waspDir options
+  (compileWarnings, appSpecOrCompileErrors) <- analyzeWaspProject waspDir options
   case appSpecOrCompileErrors of
-    Left compileErrors -> return ([], compileErrors)
+    Left compileErrors -> return (compileWarnings, toList compileErrors)
     Right appSpec ->
       case ASV.validateAppSpec appSpec of
         [] -> do
           (generatorWarnings, generatorErrors) <- Generator.writeWebAppCode appSpec outDir (sendMessage options)
-          return (map show $ generatorWarningsFilter options generatorWarnings, map show generatorErrors)
+          let filteredGeneratorWarnings = map show $ generatorWarningsFilter options generatorWarnings
+          return (filteredGeneratorWarnings ++ compileWarnings, map show generatorErrors)
         validationErrors -> do
-          return ([], map show validationErrors)
+          return (compileWarnings, map show validationErrors)
 
 analyzeWaspProject ::
   Path' Abs (Dir WaspProjectDir) ->
   CompileOptions ->
-  IO (Either [CompileError] AS.AppSpec)
+  IO ([CompileWarning], Either (NonEmpty CompileError) AS.AppSpec)
 analyzeWaspProject waspDir options = do
+  warnings <- warnIfDotEnvPresent waspDir
   maybeWaspFilePath <- findWaspFile waspDir
   case maybeWaspFilePath of
-    Nothing -> return $ Left ["Couldn't find a single *.wasp file."]
+    Nothing -> return (warnings, Left $ fromList ["Couldn't find a single *.wasp file."])
```

Comment:
Similar like above, you don't have to construct pair so many times, instead we could have this big case expression return `Either (NonEmpty Error) AppSpec` and then we at the last line combine that with `warnings` into a pair and return it.